Ports/build_all.sh

```diff
@@ -1,89 +1,121 @@
+LIGHT_WHITE='\033[1m'
+LIGHT_RED='\033[1;31m'
+LIGHT_CYAN='\033[1;36m'
+LIGHT_YELLOW='\033[1;33m'
+RESET_COLOR='\033[0m'
 
-case "$1" in
-    clean)
-        clean=true
-        ;;
-    verbose)
-        verbose=true
-        ;;
-    *)
-        ;;
-esac
+some_failed=false
+action='build'
+verbose=false
+failfast=false
 
-case "$2" in
-    clean)
-        clean=true
-        ;;
-    verbose)
-        verbose=true
-        ;;
-    *)
-        ;;
-esac
+for arg in "$@"; do
+    case "$arg" in
+        clean*)
+            action="$arg"
+            ;;
+        verbose)
+            verbose=true
+            ;;
+        failfast)
+            failfast=true
+            ;;
+    esac
+done
 
 some_failed=false
-built_ports=""
+processed_ports=()
 
-for file in *; do
-    if [ -d $file ]; then
-        pushd $file > /dev/null
-            port=$(basename $file)
-            port_built=0
-            for built_port in $built_ports; do
-                if [ "$built_port" = "$port" ]; then
-                    port_built=1
-                    break
-                fi
-            done
-            if [ $port_built -eq 1 ]; then
-                echo "Built $port."
-                popd > /dev/null
-                continue
-            fi
-            if ! [ -f package.sh ]; then
-                echo "ERROR: Skipping $port because its package.sh script is missing."
-                popd > /dev/null
-                continue
-            fi
+serenity_root=$(realpath "$(dirname "${BASH_SOURCE[0]}")/../")
+serenity_ports_dir="${serenity_root}/Ports"
+
+function do_log() {
+    [[ -z "$2" ]] && return
 
-            if [ "$clean" == true ]; then
-                if [ "$verbose" == true ]; then
-                    ./package.sh clean_all
-                else
-                    ./package.sh clean_all > /dev/null 2>&1
+    case "$1" in
+        success)
+            echo -e "${LIGHT_GREEN}[#]${RESET_COLOR} $2"
+            ;;
+        process)
+            echo -e "${LIGHT_WHITE}[~]${RESET_COLOR} $2"
+            ;;
+        info)
+            echo -e "${LIGHT_CYAN}[*]${RESET_COLOR} $2"
+            ;;
+        warn)
+            echo -e "${LIGHT_YELLOW}[!]${RESET_COLOR} $2"
+            ;;
+        error)
+            echo -e "${LIGHT_RED}[x]${RESET_COLOR} $2"
+            ;;
+    esac
+}
+
+do_build_port() {
+    do_log process "Building $port_name"
+    if $verbose; then
+        ./package.sh
+    else
+        ./package.sh &> /dev/null
+    fi
+}
+
+do_clean_port() {
+    do_log process "Cleaning $port_name"
+    if $verbose; then
+        ./package.sh "$1"
+    else
+        ./package.sh "$1" &> /dev/null
+    fi
+}
+
+for port_dir in $(find "$serenity_ports_dir" -maxdepth 1 -type d -not -path "$serenity_ports_dir" | sort); do
+    port_name="$(basename "$port_dir")"
+    if [[ " ${processed_ports[*]} " == *" $port_name "* ]]; then
+        do_log info "$port_name is already processed"
+        continue
+    fi
+    cd "$port_dir" || continue
```

Comment:
This feels like something that we should fail on rather than silently skipping. Would this even feasibly happen, other than the user going around deleting ports while the script is running?

Reply:
We can use `failfast` flag here. What do you thing?

Reply:
A directory that we just found via `find` that is suddenly not accessible sounds serious enough to stop in any case.

Reply:
Fixed!